cumulus/client/cirrus-executor/src/lib.rs

```diff
@@ -296,6 +303,49 @@ where
 			.ok_or(sp_blockchain::Error::Backend(format!("Header not found for {:?}", at)))
 	}
 
+	fn block_body(&self, at: Block::Hash) -> Result<Vec<Block::Extrinsic>, sp_blockchain::Error> {
+		self.client
+			.block_body(&BlockId::Hash(at))?
+			.ok_or(sp_blockchain::Error::Backend(format!("Block body not found for {:?}", at)))
+	}
+
+	fn create_extrinsic_execution_proof(
+		&self,
+		extrinsic_index: usize,
+		parent_header: &Block::Header,
+		current_hash: Block::Hash,
+	) -> Result<StorageProof, GossipMessageError> {
+		let extrinsics = self.block_body(current_hash)?;
+
+		let encoded_extrinsic = extrinsics[extrinsic_index].encode();
+
+		let block_builder = BlockBuilder::with_extrinsics(
+			&*self.client,
+			parent_header.hash(),
+			*parent_header.number(),
+			RecordProof::No,
+			Default::default(),
+			&*self.backend,
+			extrinsics,
+		)?;
+		let storage_changes = block_builder.prepare_storage_changes_before(extrinsic_index)?;
+
+		let delta = storage_changes.transaction;
+		let post_delta_root = storage_changes.transaction_storage_root;
+
+		let execution_proof = cirrus_fraud_proof::prove_execution(
+			&self.backend,
+			&*self.code_executor,
+			self.spawner.clone() as Box<dyn SpawnNamed>,
+			&BlockId::Hash(parent_header.hash()),
+			"BlockBuilder_apply_extrinsic",
```

Comment:
I'm why can't we use runtime API of the client (which we have just a few lines above) instead of manually calling these methods (there are a few places like this).

It would be extremely easy to break such stringy API without any help from the compiler, which bothers me a lot.

Reply:
Hm, what about proposing extension to runtime API that will make it possible to achieve what we want with a special runtime API call that uses some custom state? I'm fine merging this as is, but it should have a TODO in every place we have something like this and communicate with Substrate folks on how to handle it nicely.

Reply:
TODO added, will post a discussion in Substrate later.